src/rust/aggregator/src/data/user_history.rs

```diff
@@ -0,0 +1,411 @@
+                    AND place_limit_order_events.event_idx = aggregated_events.event_idx
+                )
+            "#,
+        )
+        .fetch_all(&self.pool)
+        .await
+        .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        let market_events = sqlx::query!(
+            r#"
+                SELECT * FROM place_market_order_events
+                WHERE NOT EXISTS (
+                    SELECT * FROM aggregator.aggregated_events
+                    WHERE place_market_order_events.txn_version = aggregated_events.txn_version
+                    AND place_market_order_events.event_idx = aggregated_events.event_idx
+                )
+            "#,
+        )
+        .fetch_all(&self.pool)
+        .await
+        .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        let swap_events = sqlx::query!(
+            r#"
+                SELECT * FROM place_swap_order_events
+                WHERE NOT EXISTS (
+                    SELECT * FROM aggregator.aggregated_events
+                    WHERE place_swap_order_events.txn_version = aggregated_events.txn_version
+                    AND place_swap_order_events.event_idx = aggregated_events.event_idx
+                )
+            "#,
+        )
+        .fetch_all(&self.pool)
+        .await
+        .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        for x in &limit_events {
+            sqlx::query!(
+                r#"
+                    INSERT INTO aggregator.user_history_limit VALUES (
+                        $1, $2, $3, $4, $5, $6, $7
+                    );
+                "#,
+                x.market_id,
+                x.order_id,
+                x.user,
+                x.custodian_id,
+                x.side,
+                x.self_match_behavior,
+                x.restriction,
+            )
+            .execute(&self.pool)
+            .await
+            .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+            sqlx::query!(
+                r#"
+                    INSERT INTO aggregator.user_history VALUES (
+                        $1, $2, $3, $4, $5, $6, $7, $8, $9
+                    );
+                "#,
+                x.market_id,
+                x.order_id,
+                x.time,
+                None as Option<DateTime<Utc>>,
+                x.integrator,
+                BigDecimal::zero(),
+                x.initial_size,
+                OrderStatus::Open as OrderStatus,
+                OrderType::Limit as OrderType,
+            )
+            .execute(&self.pool)
+            .await
+            .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        }
+        if !limit_events.is_empty() {
+            for x in limit_events {
+                sqlx::query!(
+                    r#"
+                        INSERT INTO aggregator.aggregated_events VALUES (
```

Comment:
Something tells me we should be using a SQL transaction here, because we might insert half the events to user history before erroring out. Then it won't insert any events into this table and we'll get duplicated work.

Reply:
Yes, just did that :+1: